File: go/binary_search.go
package main

import "fmt"

func binarySearch(arr []int, target int) int {
	low := 0
	high := len(arr) - 1

	for low <= high {
		mid := low + (high-low)/2 // Avoids overflow
		
		if arr[mid] == target {
			return mid
		} else if arr[mid] < target {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}
	return -1 // Not found
}

// Example
func main() {
	sortedArr := []int{2, 5, 8, 12, 16, 23, 38, 56, 72, 91}
	t1 := 23
	t2 := 10

	index1 := binarySearch(sortedArr, t1)
	index2 := binarySearch(sortedArr, t2)

	fmt.Printf("Searching for %d in array: %v\n", t1, sortedArr)
	fmt.Printf("Found %d at index: %d\n", t1, index1)
	
	fmt.Printf("Searching for %d in array: %v\n", t2, sortedArr)
	fmt.Printf("Found %d at index: %d\n", t2, index2)
}

File: rust/binary_search.rs
fn binary_search(arr: &[i32], target: i32) -> Option<usize> {
    let mut low = 0;
    let mut high = arr.len();

    while low < high {
        let mid = low + (high - low) / 2;
        if arr[mid] == target {
            return Some(mid);
        } else if arr[mid] < target {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    None
}

fn main() {
    let arr = [1, 3, 5, 7, 9, 11, 13];
    let target = 7;

    match binary_search(&arr, target) {
        Some(index) => println!("Found {} at index {}", target, index),
        None => println!("{} not found in array", target),
    }
}
